src/all-blend.ts

```diff
@@ -1,26 +1,20 @@
-import { ifThen, defn, int, eq, ret } from "@thi.ng/shader-ast";
+import { defn, eq, ifThen, int, ret } from "@thi.ng/shader-ast";
+import type { BlendMode, BlendModeDef, Color } from "./api";
 import { BLEND_MODES_3, BLEND_MODES_4 } from "./constants";
 
-export const blendModeSelect3 = defn(
-  "vec3",
-  "blendModePick",
-  ["int", "vec3", "vec3", "float"],
-  (mode, base, blend, opacity) => {
-    const select = Object.values(BLEND_MODES_3).map((blendFn, i) =>
-      ifThen(eq(mode, int(i)), [ret(blendFn(base, blend, opacity))])
-    );
-    return [...select];
-  }
-);
+const blendModeSelect = <T extends Color>(
+  type: T,
+  modes: Record<BlendMode, BlendModeDef<T>>
+) =>
+  defn(
+    type,
+    "blendModePick",
+    ["int", type, type, "float"],
+    (mode, base, blend, opacity) =>
+      Object.values(modes).map((blendFn, i) =>
+        ifThen(eq(mode, int(i)), [ret(blendFn(base, blend, opacity))])
+      )
+  );
```

Comment:
Being able to use these kinds of higher order functions is my fave feature in shader-ast. This can remove so much bloat & improve re-use...

Reply:
Yeah I was pretty impressed that this just worked first time..